### PR TITLE
Tags: Add project tags, negation, `dev`/`autodocs`/`test` system tags

### LINKED ISSUE #9209: Add an option to hide a story from the sidebar
**Is your feature request related to a problem? Please describe.**
I am doing visual regression testing with Storybook iframes. I have some docs pages that aren't specific for a single component, for example, I have an "icons" page that shows all the icons of the library in a grid. I would like to test all these icons but since they are on a docs page, I don't have the iframe available.

**Describe the solution you'd like**
I would like to be able to create stories that are hidden on the sidebar but can still be accessible by the URL. Perhaps a per-story configuration.

**Describe alternatives you've considered**
I can create a story just for the icons grid, but it is redundant and a bit out of place since I already have the icons doc page that has a lot more info

**Are you able to assist bring the feature to reality?**
Yes, if the suggestion makes sense for the project, I can find some time to implement it

### LINKED ISSUE #21085: [Feature Request]: Storybook 7 - opt-out of autodocs for a single story when `autodocs: true`
### Is your feature request related to a problem? Please describe

When using `autodocs: true` in `main.js` a doc is generated for every story. But there may be edge-cases like for a `Welcome` page where we do not want a doc page generated.

### Describe the solution you'd like

 It would be nice to be able to opt-out of the autodocs on a story level with:

```tsx
export default {
  title: 'Welcome to Storybook',
  component: () => <>Welcome</>,
  parameters: {
    docs: {
      autodocs: false, // overrides the default just for this story
    },
  },
} as Meta;
```

### Describe alternatives you've considered

Using `autodocs: 'tags'` but this would be a lot of boilerplate since all our stories would require the `autodocs` tag, and only a single `Welcome` story would be opted out. 

### Are you able to assist to bring the feature to reality?

yes, I can

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shilman**
Closes #21085 
Closes #9209

Based on https://github.com/ComponentDriven/csf/pull/80

### Project tags

Added project tags:

```js
// .storybook/preview.js
export default {
  tags: ['autodocs'],
}
```

This means that we don't need the `docs.autodocs` configuration anymore.
1. To enable autodocs globally, add `autodocs` tag to `preview.js`
2. To opt-out of autodocs on individual components, add `!autodocs` to the story meta
3. To make autodocs opt-in on a per-component basis, don't add `autodocs` to preview.js and selectively add `autodocs` to the components you want autodocs for.

### Tag negation

Add the ability to remove tags that have been added higher up in the hierarchy using `!tag` syntax:

```js
// Button.stories.js

export default {
  component: Button,
  tags: ['foo', 'bar', '!autodocs']
}

export const Primary = {
  tags: ['!foo', 'baz']
}
```

### Added dev/docs/test system tags

Added new "system" tags `dev` and `test` that automatically get applied to every store (and can be removed with `!dev` etc.
- `!dev` removes a story/component from the sidebar
- `!autodocs` removes a story from autodocs
- `!test` removes a story from test runs

### Removed story/docs system tags

Storybook was automatically adding the tag `story` to stories and `docs` to docs nodes (autodocs, mdx, etc.).

I've removed these, since the use case wasn't clear and since `docs` now means something else per above ☝️ 

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Run a sandbox or test project with the new version
2. Verify legacy `docs.autodocs` settings in `main.js`
  a. That they still work as before
  b. That they give a deprecation warning in the browser console and that it makes sense.
3. Remove the `docs.autodocs` setting and restart
4. Test the following settings for 'autodocs':
  a. no `autodocs` tag is added to `.storybook/preview.js`
  b. `autodocs` tag is added to a component (**opt-in case**)
  c. `autodocs` tag is added to `.storybook/preview.js`
  d. `!autodocs` tag is added to a component (**opt-out case**)
  d. `!autodocs` tag is added to a story (**opt-out case** of a story in autodocs)
5. Test the same settings for the `dev` tag. Note that the `dev` tag is added globally by default.
  a. Disable `dev` tag globally by adding `!dev` to `.storybook/preview.ts` (**opt-in case**)
  b. Add `dev` tag to a component and verify that it appears in the sidebar
  c. Add `!dev` tag to a story within that component and verify that it disappears from the sidebar
  d. Enable `dev` globally by removing `!dev` from `.storybook/preview.js`
  e. Add `!dev` to a component and verify that it disappears from the sidebar (**opt-out case**)
7. Run the `test-runner` with `--includeTags=test` and verify that only those stories get tested

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: code/lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 1]
```diff
@@ -312,7 +314,9 @@ export class StoryIndexGenerator {
     const hasAutodocsTag = entries.some((entry) => entry.tags.includes(AUTODOCS_TAG));
     const isStoriesMdx = entries.some((entry) => entry.tags.includes(STORIES_MDX_TAG));
     const createDocEntry =
-      autodocs === true || (autodocs === 'tag' && hasAutodocsTag) || isStoriesMdx;
+      (autodocs === true && !entries.some((entry) => entry.tags.includes(NO_AUTODOCS_TAG))) ||
```
Comment (tmeasday) [inline code/lib/core-server/src/utils/StoryIndexGenerator.ts:317]: Makes sense to pull this out to `hasNoAutodocsTag`, and then `hasAutodocsTag = !hasNoAutodocs & ...`, or even to call `combineTags` on the tags first.

File: code/lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 2]
```diff
@@ -326,7 +330,12 @@ export class StoryIndexGenerator {
         name,
         importPath,
         type: 'docs',
-        tags: [...metaTags, 'docs', ...(!hasAutodocsTag && !isStoriesMdx ? [AUTODOCS_TAG] : [])],
+        metaTags,
+        tags: combineTags(
+          ...metaTags,
+          'docs',
+          ...(!hasAutodocsTag && !isStoriesMdx ? [AUTODOCS_TAG] : [])
```
Comment (tmeasday) [inline code/lib/core-server/src/utils/StoryIndexGenerator.ts:337]: Could we make `combineTags` drop duplicates so we don't need this check for `hasAutodocsTag`?

**Review by tmeasday — APPROVED** [review]
Again, LGTM, but need to see the CSF PR

File: code/lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 3]
```diff
@@ -307,29 +317,27 @@ export class StoryIndexGenerator {
...
-      autodocs === true || (autodocs === 'tag' && hasAutodocsTag) || isStoriesMdx;
+    const createDocEntry = (hasAutodocsTag && this.options.docs.autodocs !== false) || isStoriesMdx;
 
     if (createDocEntry && this.options.build?.test?.disableAutoDocs !== true) {
       const name = this.options.docs.defaultName ?? 'Docs';
       const { metaId } = indexInputs[0];
       const { title } = entries[0];
-      const metaTags = indexInputs[0].metaTags || [];
       const id = toId(metaId ?? title, name);
+      const tags = combineTags(...projectTags, ...(indexInputs[0].tags ?? []));
```
Comment (JReinhold) [inline code/lib/core-server/src/utils/StoryIndexGenerator.ts:332]: I think this reverses the fix you did in https://github.com/storybookjs/storybook/pull/25537

With this change, the autodocs entry will again inherit all the tags from the primary story.

I'm not sure there's a clean workaround for it though.

**Pull request merged**